src/wavm/openmp.cpp

```diff
@@ -452,147 +456,242 @@ WAVM_DEFINE_INTRINSIC_FUNCTION(env,
-                               I32 keyPtr,
-                               I64 value)
+// -------------------------------------------------------
+// FOR LOOP STATIC INIT
+// -------------------------------------------------------
+
+enum sched_type : int
 {
-    const std::shared_ptr<spdlog::logger>& logger = faabric::util::getLogger();
-    logger->debug("S - __faasmp_incryby {} {}", keyPtr, value);
+    sch_lower = 32, /**< lower bound for unordered values */
+    sch_static_chunked = 33,
+    sch_static = 34, /**< static unspecialized */
+};
 
-    Runtime::Memory* memoryPtr = getExecutingWAVMModule()->defaultMemory;
-    std::string key{ &Runtime::memoryRef<char>(memoryPtr, (Uptr)keyPtr) };
-    faabric::redis::Redis& redis = faabric::redis::Redis::getState();
-    return redis.incrByLong(key, value);
-}
+template<typename T>
+void for_static_init(I32 schedule,
+                     I32* lastIter,
+                     T* lower,
+                     T* upper,
+                     T* stride,
+                     T incr,
+                     T chunk);
 
-WAVM_DEFINE_INTRINSIC_FUNCTION(env,
-                               "__faasmp_getLong",
-                               I64,
-                               __faasmp_getLong,
-                               I32 keyPtr)
+template<typename T>
+void for_static_init(I32 schedule,
+                     I32* lastIter,
+                     T* lower,
+                     T* upper,
+                     T* stride,
+                     T incr,
+                     T chunk)
 {
-    const std::shared_ptr<spdlog::logger>& logger = faabric::util::getLogger();
-    logger->debug("S - __faasmp_getLong {}", keyPtr);
+    // Unsigned version of the given template parameter
+    typedef typename std::make_unsigned<T>::type UT;
 
-    Runtime::Memory* memoryPtr = getExecutingWAVMModule()->defaultMemory;
-    std::string key{ &Runtime::memoryRef<char>(memoryPtr, (Uptr)keyPtr) };
-    faabric::redis::Redis& redis = faabric::redis::Redis::getState();
-    return redis.getLong(key);
-}
+    auto logger = faabric::util::getLogger();
+    auto ctx = getOpenMPContext();
 
-/**
- * This function is just around to debug issues with threaded access to stacks.
- */
-WAVM_DEFINE_INTRINSIC_FUNCTION(env,
-                               "__faasmp_debug_copy",
-                               void,
-                               __faasmp_debug_copy,
-                               I32 src,
-                               I32 dest)
-{
-    const std::shared_ptr<spdlog::logger>& logger = faabric::util::getLogger();
-    logger->debug("S - __faasmp_debug_copy {} {}", src, dest);
+    if (ctx.level->numThreads == 1) {
+        *lastIter = true;
 
-    // Get pointers on host to both src and dest
-    Runtime::Memory* memoryPtr = getExecutingWAVMModule()->defaultMemory;
-    int* hostSrc = &Runtime::memoryRef<int>(memoryPtr, src);
-    int* hostDest = &Runtime::memoryRef<int>(memoryPtr, dest);
+        if (incr > 0) {
```

Comment:
This chunk of code is quite complex, but it's just been moved from higher up the file and is otherwise unchanged.